expression/builtin_time.go

```diff
@@ -3342,6 +3354,22 @@ func (c *addDateFunctionClass) getFunction(ctx sessionctx.Context, args []Expres
 			args[0] = BuildCastFunction(ctx, args[0], tp)
 		}
 		bf.setDecimalAndFlenForDatetime(int(types.MaxFsp))
+
+		if dateEvalTp == types.ETString {
+			bf.tp.Tp = mysql.TypeString
+		}
+
+		if dateEvalTp == types.ETDatetime && args[0].GetType().Tp == mysql.TypeDate {
+			switch strings.ToUpper(unit) {
+			// If the unit is YMD, the return type is date.
+			case "YEAR", "MONTH", "DAY", "YEAR_MONTH":
+				tp := types.NewFieldType(mysql.TypeDate)
+				tp.Decimal = 0
```

Comment:
Use setDecimalAndFlenForDate() to set them.

Reply:
done